### PR TITLE
Cruft PR workflow

### LINKED ISSUE #138: Template sync fails when changes are made to workflow files
In a test-run, the template sync failed since apparently
the `sync.yaml` action is not permitted to make changes to files in the
`.github/workflows` directory:
https://github.com/icbi-lab/infercnvpy/actions/runs/3875431171/jobs/6607954535

This is a bit of a problem for which we still need to figure out a
solution. One of them is to use a personal access token. Another one
could be to centralize template sync, which I would have liked to avoid.

In #137, I roll back the changes made to workflow files since the last release,
as a short-term fix.

### PR DESCRIPTION
**Pull request by flying-sheep**
Derived from https://gist.github.com/grst/7593e7ff965df9a6aeab5c4fbdd796bf

Close #138

### CODE REVIEW

File: .github/workflows/cruft-prs.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
+name: make cruft PRs for all projects using us
+on:
+  release:
```
Comment (grst) [inline .github/workflows/cruft-prs.yml:3]: I'd also add workflow_dispatch

Comment (flying-sheep) [reply .github/workflows/cruft-prs.yml:3]: Hmm, then we’d have to add all the release metadata used in the PR title body as `input`s

Comment (flying-sheep) [reply .github/workflows/cruft-prs.yml:3]: OK, with the script it’s easy

**Comment by flying-sheep** [comment]
blocked by https://github.com/charliermarsh/ruff/issues/4134

File: scripts/src/scverse_template_scripts/cruft_prs.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    return True
+
+
+def make_pr(con: GitHubConnection, release: GHRelease, repo_url: str) -> None:
+    pr = PR(release)
+    log.info(f"Sending PR to {repo_url}: {pr.title}")
+
+    # create fork, populate branch, do PR from it
+    origin = con.gh.get_repo(repo_url.removeprefix("https://github.com/"))
+    repo = origin.create_fork()
```
Comment (grst) [inline scripts/src/scverse_template_scripts/cruft_prs.py:142]: does this work if the fork already exists? 

Do we need to pull upstream into the fork if it already exists from the last update?

File: scripts/src/scverse_template_scripts/cruft_prs.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    with TemporaryDirectory() as td:
+        updated = cruft_update(con, repo, Path(td), pr)
+    if updated:
+        origin.create_pull(pr.title, pr.body, ...)  # TODO
```
Comment (grst) [inline scripts/src/scverse_template_scripts/cruft_prs.py:146]: what happens if a pull request already exists from the last update but hasn't been merged yet?

File: scripts/src/scverse_template_scripts/cruft_prs.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        return f"Update template to {self.release.tag_name}"
+
+    @property
+    def branch(self) -> str:
+        return f"template-update-{self.release.tag_name}"
+
+    @property
+    def body(self) -> str:
+        template_usage = "https://cookiecutter-scverse-instance.readthedocs.io/en/latest/template_usage.html"
+        return f"""\
```
[the remark this answers is not in the document]
Comment (flying-sheep) [reply scripts/src/scverse_template_scripts/cruft_prs.py:63]: since we insert the body, we need to `indent` that one. and once we use that, a refactor could break things as the `indent`, would need a different argument. I’d rather use a template file than that.

File: scripts/src/scverse_template_scripts/cruft_prs.py (hunk 2 above)
Comment (flying-sheep) [reply scripts/src/scverse_template_scripts/cruft_prs.py:146]: great point, we should add a test for that

File: scripts/src/scverse_template_scripts/cruft_prs.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (grst) [reply scripts/src/scverse_template_scripts/cruft_prs.py:63]: ah got it!

File: scripts/src/scverse_template_scripts/cruft_prs.py (hunk 2 above)
Comment (flying-sheep) [reply scripts/src/scverse_template_scripts/cruft_prs.py]: hm, testing the GitHub API doesn’t seem to be possible without mocking everything manually

File: scripts/src/scverse_template_scripts/cruft_prs.py (hunk 3 above)
Comment (flying-sheep) [reply scripts/src/scverse_template_scripts/cruft_prs.py]: done

File: scripts/pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+[tool.hatch.envs.test.scripts]
+test = "pytest"
+
+[tool.pytest.ini_options]
+addopts = [
+    "--import-mode=importlib",
+    "-pgit",
+    "-pnetwork", "--disable-network",
+    "-presponsemock"
+]
```
Comment (grst) [inline scripts/pyproject.toml:57]: why is this needed?

File: scripts/pyproject.toml [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    "S",
+    "T",
+    "TID",
+    "UP",
+    "W",
+    "YTT",
+]
+ignore = [
+    "S101",  # assert should be allowed
+    "S603",  # suprocess with shell=False should be allowed
```
Comment (grst) [inline scripts/pyproject.toml:97]:
```suggestion
    "S603",  # subprocess with shell=False should be allowed
```

**Review by grst — APPROVED** [review]
Looks beautiful! 
Now how do we best go about merging this and testing this in a real world setting? 

In general, I would suggest that the next release of the cookiecutter template *only* includes the changes from this PR (i.e. delete `sync.yaml`). This makes it easier to communicate to the template users that we switched to a new system and the respective PRs will be easy to review and merge (there shouldn't be any merge conflicts).

File: scripts/pyproject.toml (hunk 5 above)
Comment (flying-sheep) [reply scripts/pyproject.toml:57]: The `-p???` switches activate the pytest plugins used and make pytest complain if they’re missing. The `--disable-network` switch makes sure all network requests are mocked out.

**Comment by grst** [comment]
merge this PR and cherry-pick it in the `v0.2.x` branch, then make a release and see what happens?

Or is there a better way  of testing your "on release" action before actually making a release?

**Comment by flying-sheep** [comment]
You’re right!
